flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/SharedObjects.java

```diff
@@ -0,0 +1,156 @@
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * This rule allows objects to be used both in the main test case as well as in UDFs by using
+ * serializable {@link SharedReference}s. Usage:
+ *
+ * <pre><code>
+ * {@literal    @Rule}
+ *     public final SharedObjects sharedObjects = SharedObjects.create();
+ *
+ * {@literal    @Test}
+ *     public void test() throws Exception {
+ *         StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+ *         SerialRef<Queue<Long>> list = sharedObjects.add(new ConcurrentLinkedQueue<>());
+ *         int n = 10000;
+ *         env.setParallelism(100);
+ *         env.fromSequence(0, n).map(i -> list.get().add(i));
+ *         env.execute();
+ *         assertEquals(n + 1, list.get().size());
+ *         assertEquals(
+ *                 LongStream.rangeClosed(0, n).boxed().collect(Collectors.toList()),
+ *                 list.get().stream().sorted().collect(Collectors.toList()));
+ *     }
+ * </code></pre>
+ *
+ * <p>Note that since the shared objects are accessed through multiple threads, they need to be thread-safe or accessed in a thread-safe manner.
+ */
+public class SharedObjects extends ExternalResource implements Serializable {
+    /**
+     * Instance-cache used to make a SharedObjects accessible for multiple threads.
+     */
+    private static final Map<Integer, SharedObjects> INSTANCES = new ConcurrentHashMap<>();
+    private static final AtomicInteger LAST_ID = new AtomicInteger();
+    /**
+     * Identifier of the SharedObjects used to retrieve the original instance during deserialization.
+     */
+    private final int id;
+    /**
+     * All registered objects for the current test case. The objects are purged upon completion.
+     */
+    private final transient Map<SharedReference<?>, Object> objects = new ConcurrentHashMap<>();
+
+    private SharedObjects(int id) {
+        this.id = id;
+    }
+
+    /**
+     * Creates a new instance. Usually that should be done inside a JUnit test class as an
+     * instance-field annotated with {@link org.junit.Rule}.
+     */
+    public static SharedObjects create() {
+        return new SharedObjects(LAST_ID.getAndIncrement());
+    }
+
+    /**
+     * Adds a new object to this {@code SharedObjects}. Although not necessary, it is recommended
+     * to only access the object through the returned {@link SharedReference}.
+     */
+    public <T> SharedReference<T> add(T object) {
+        SharedReference<T> tag = new DefaultTag<>(this, objects.size());
+        objects.put(tag, object);
+        return tag;
+    }
+
+    @Override
+    protected void before() throws Throwable {
+        INSTANCES.put(id, this);
+    }
+
+    @Override
+    protected void after() {
+        objects.clear();
+        INSTANCES.remove(id);
+    }
+
+    private Object readResolve() throws ObjectStreamException {
```

Comment:
can we put this somewhere more prominent, because when scanning the file this is easily missed yet is fundamental to the functionality.

Reply:
I'd treat that as an implementation detail to be honest. Maybe it's rather confusing that SharedObjects is also serializable and we should remove that and just do all the resolution in `SharedReference#get`? I'll add a proposal.